### PR TITLE
EES-1077 Use sequentially generated Guid's

### PR DESCRIPTION
**Pull request by benoutram**
By convention, primary keys of type Guid are set up to have values generated for inserted entities using `Microsoft.EntityFrameworkCore.ValueGeneration.SequentialGuidValueGenerator` which generates sequential values using the same algorithm as `NEWSEQUENTIALID()` in Microsoft SQL Server.

There are some cases in our application where `Id = Guid.NewGuid` has been set prior to adding the entity e.g. in `CreatePublicationAsync`. 

These are generating non-sequential Guid's which are impacting the clustered PK indexes, causing fragmentation. Each row of data is inserted into the middle of the clustered index (out of sequential order) causing SQL Server to make room for the data by rearranging the cluster.

Where the Id does not need to be known prior to adding, they have been removed to force generation using the default value generator when the entity is added.

Where the Id does need to be known prior to adding, they have been replaced by the value from `SequentialValueGenerator.NewGuid`. This is a copy of the EF Core value generator without the `ValueGenerator` interface so it can be used without the database entity.

With this PR for example, we can see that for an imported Subject, Observations ordered by Id are in order of the Csv Row Id that they were originally generated from.  Prior to this because the Observation Id was `Id = Guid.NewGuid` the order was completely random.

Values created by different machines or different function instances will create gaps but should still be unique and sequential.

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServicePermissionTests.cs [hunk 1]
```diff
@@ -179,7 +180,7 @@ public async void GetMyReleasesForReleaseStatusesAsync_CanViewAllReleases()
             
             var service = new ReleaseService(contentDbContext.Object, AdminMapper(),
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServicePermissionTests.cs:181]: Seems like we could refactor out a `createReleaseService` method for all of these tests.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs [hunk 2]
```diff
@@ -49,7 +50,7 @@ public void CreateReleaseNoTemplate()
             {
                 var releaseService = new ReleaseService(context, AdminMapper(),
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs:51]: Could refactor out a `createReleaseService` method.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs [hunk 3]
```diff
@@ -48,23 +48,25 @@ public class ReleaseService : IReleaseService
...
-            IFileStorageService fileStorageService, 
+            ITableStorageService coreTableStorageService,
+            IFileStorageService fileStorageService,
             IImportStatusService importStatusService,
-	        IFootnoteService footnoteService, 
+            IFootnoteService footnoteService,
             StatisticsDbContext statisticsDbContext,
-            IDataBlockService dataBlockService)
+            IDataBlockService dataBlockService, 
+            IGuidGenerator guidGenerator)
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs:69]: Not something we need to do for this ticket, but this class is definitely feeling like it's becoming harder to maintain with its continuously growing dependency list. Feels like we should act on the comment above at some point soon.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ImportService.cs [hunk 4]
```diff
@@ -103,7 +105,7 @@ private ImportMessage BuildMessage(string dataFileName, Guid releaseId, int numR
             
             return new ImportMessage
             {
-                SubjectId = Guid.NewGuid(),
+                SubjectId = _guidGenerator.NewGuid(),
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ImportService.cs:108]: For smaller tables probably not but primary keys of type Guid are set up to have values generated for inserted entities using Microsoft.EntityFrameworkCore.ValueGeneration.SequentialGuidValueGenerator so hopefully this is only generating a new guid in the same way that EF core would be doing if we were to save a Subject with no Id field set.

**Comment by benoutram** [comment]
> Is there any plan for how we're going to deal with existing data not using sequential GUIDs?

@ntsim I had thought that existing data should not be a problem once the indexes are rebuilt. This has been done on Dev now and there's a report of that showing the decrease in storage size on EES-1077.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs (hunk 3 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs:69]: There's already a ticket for this, EES-212. I will update the reference.

**Comment by ntsim** [comment]
> > Is there any plan for how we're going to deal with existing data not using sequential GUIDs?
> 
> @ntsim I had thought that existing data should not be a problem once the indexes are rebuilt. This has been done on Dev now and there's a report of that showing the decrease in storage size on EES-1077.

Awesome, sounds like its not going to be a problem then!

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServicePermissionTests.cs (hunk 1 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServicePermissionTests.cs]: Done

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs (hunk 2 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs]: Done

**Pull request merged**